### DISCUSSION TITLE
Username lettercase in the new store

### DISCUSSION THREAD

**Discussion by vramik** [keycloak/keycloak#10198] [Archive]
looking at the code it seems we always store user's username (and also email) in lowered-case format: https://github.com/keycloak/keycloak/blob/main/model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java#L110 and https://github.com/keycloak/keycloak/blob/main/model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java#L337 even though it's not specified in java-doc. 

Then when looking up for a user by the given username it's specified by java-doc that the username should be: `@param username case insensitive username (case-sensitivity is controlled by storage)`
- `JpaUserProvider `uses named query `getRealmUserByUsername`: `select u from UserEntity u where u.username = :username and u.realmId = :realmId` and param username is set by `query.setParameter("username", username.toLowerCase())` ... this query is supported by index
- `MapUserProvider` uses MCB where operator `ILIKE` is used: https://github.com/keycloak/keycloak/blob/main/model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java#L511 ... if it'd be implemented using `ILIKE` operator the index couldn't be used.

Question: Could it be safely assumed that username is always stored in lowered-case format? Or is that implementation specific?

Based on that we should decide if we could:
1. change `MapUserProvider` to use operator `EQ`: `.compare(SearchableFields.USERNAME, Operator.EQ, username.toLowerCase());`
2. edit java-doc for `UserProvider.addUser` specifying that username should be stored as lower-case
3. edit java-doc for `UserLookupProvider.getUserByUsername` saying it is **exact** case-insensitive search

note: it probably applies to email as well, but `MapStorageProvider.getUserByEmail` is implemented following way: `.compare(SearchableFields.EMAIL, Operator.EQ, email);` - there is no `email.toLowerCase()`, while `JpaUserProvider` uses `email.toLowerCase()`.

---
EDIT: Related issues:
- #8893
- #10245

**Comment by hmlnarik** [comment]
> Question: Could it be safely assumed that username is always stored in lowered-case format? Or is that implementation specific?

No, in the new storage it should be possible to keep letter case.

The current behaviour does not cover some usecases and we already had RFEs for keeping the letter case as provided:
- https://issues.redhat.com/browse/KEYCLOAK-8456
- https://issues.redhat.com/browse/KEYCLOAK-5699
- https://issues.redhat.com/browse/KEYCLOAK-2858

These RFEs were rejected by the time of their triage since the decision was based on legacy implementation of the storage. The new store may behave differently.

I believe the user storage map provider should specify an option whether username letter case is significant or not. If not, let's search by lowercase / uppercase (could be an option of the provider). The storage implementation might lack the feature support for either of these - for example for external sources (say LDAP or REST) it cannot arbitrarily set whether they do or do not support letter case resolution. Keycloak has to follow their settings.

**Reply by vramik** [reply]
> I believe the user storage map provider should specify an option whether username letter case is significant or not.

reading this it seems we should introduce new method variants in `UserProvider` and `UserLookupProvider` with default implementation keeping behavior of legacy store, while `MapStorageProvider` would override the default implementation allowing to specify letter case (either for username of email).

Perhaps configuration option defaulting to current behavior could be introduced in map users spi. 

WDYT?

**Reply by hmlnarik** [reply]
> reading this it seems we should introduce new method variants in `UserProvider` and `UserLookupProvider` with default implementation keeping behavior of legacy store, while `MapStorageProvider` would override the default implementation allowing to specify letter case (either for username of email).

It needs to be stressed in Javadoc and later the documentation that the lettercase significance depends on storage, but at this moment I don't see the need to introduce any new method variants in those providers.

> Perhaps configuration option defaulting to current behavior could be introduced in map users spi.

Just to be clear - this option should be introduced on the level of [map storage user provider factory](https://github.com/keycloak/keycloak/blob/main/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java) and applied in appropriate [criteria builder](https://github.com/keycloak/keycloak/blob/90d4e586b6ce92c0b4211586ad554a5a07566417/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/group/JpaGroupModelCriteriaBuilder.java#L56), not [map users provider factory](model/map/src/main/java/org/keycloak/models/map/user/MapUserProviderFactory.java), so that it is possible to configure the behaviour per realm.

**Reply by vramik** [reply]
Currently `MapUserProvider` prevents storing username with the letter case significance by https://github.com/keycloak/keycloak/blob/main/model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java#L337

I've created an issue for enabling it: https://github.com/keycloak/keycloak/issues/10245

As part of the issue should be also update of current implementations of `MapStorageProviderFactory`, such as `ConcurrentHashMapStorageProviderFactory`, `HotRodMapStorageProviderFactory` and `JpaMapStorageProviderFactory`, where the option to specify lettecase significance for username would be taken from config.

**Reply by mhajas** [reply]
Is there some advantage of having the letter case of stored username configurable? To me, it seems the most important part of this is to have uniqueness check and searching by username case-insensitive (this should be true all the time regardless of how the data are stored). How the username is stored is given by the storage capabilities whether it is able to store things with preserving letter case.

**Reply by hmlnarik** [reply]
Thank you.

The lettercase should be consistent across the stores in the tree. If a map store "above" is not capable of differentiating the letter case, the store "below" should behave as if it was not capable of the same even if technically it it did.

As for "searching by username case-insensitive": this is the feature that is considered a bug by some, see the referenced issues above. Thus we need a configurability on whether searching by username is lettercase sensitive or not.

**Comment by vramik** [comment]
related issue: https://github.com/keycloak/keycloak/issues/8893

**Comment by stianst** [comment]
I agree that the store in the future should keep the lettercase, then we rather solve it when looking up user based on username. I can see folks wanting both to solve the problem with LDAP/saml/etc and also still perhaps wanting to just type username with lowercase perhaps. Any ways okay with the change in general as long as current (oddities) don't end up being blockers or breaking behaviour.

**Comment by vramik** [comment]
On the last team call we've discussed that the best solution, to allow storing username in case-sensitive format and also to be able to support fast search when searching in case-insensitive mode, would be to store both variants of the username (both `username` and `usernameLowerCase`) for all implementations of the user map storage.

This decision implies that we might enhance the `MapUserEntity` with `usernameLowerCase` field. If we do so, it seems we could be able to introduce new searchable field `USERNAME_CASE_SENSITIVE`. 

Then the `USERNAME_CASE_SENSITIVE` could be used directly in `MapUserProvider` and based on the setting (username-case-sensitive-search vs. username-case-insensitive-search) decide to use `USERNAME_CASE_SENSITIVE` + `username` field OR `USERNAME` + `usernameLowerCase` in the search.

There would be some stores, added in a future, which won't allow to distinguish letter case. Those stores should throw the `CriterionNotSupportedException` in case some unsupported search would be done. 

Tree store then should manage the setting for the whole tree and ensure it is set correctly.

**Comment by mhajas** [comment]
@vramik Question regarding the usernameSensitivity switch. Should we take it into account only in the method `getUserByUsername` or also in methods where we search by username [1,2]? 

Example: given we have storage configured as case sensitive, and we have the following users `admin`, `Admin`, `ADmin`. Now, for sure when using the method `getUserByUsername("admin")` we want to always get only the first user. However, when doing a search `searchForUserStream(realm, ["username" -> "admin"])` do we want to return all users in the stream or only the first one? 

/cc @hmlnarik 

[1] https://github.com/keycloak/keycloak/blob/main/model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java#L609
[2] https://github.com/keycloak/keycloak/blob/main/model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java#L787

**Reply by vramik** [reply]
I think we should keep the current behavior as default new one. As agreed the default should be `case-insensitive`. 

That being said I've changed it to use `mcb.compare(SearchableFields.USERNAME, Operator.LIKE, value.toLowerCase())` for `case-insensitive` search and `mcb.compare(SearchableFields.USERNAME_CASE_SENSITIVE, Operator.LIKE, value)` for the `case-sensitive` one. 

It can be changed in case of a need, of course.

**Reply by mhajas** [reply]
Sure, we should keep the default behaviour as is now. So you plan to change also `searchForUserStream` to take the switch into account?

**Reply by vramik** [reply]
I've changed it this way in the PR. If there will be a requirement for changing it, I'll update.

**Comment by vramik** [comment]
**Heterogeneous cluster**
During migration to new version there will be two adjacent versions of keycloak within one cluster. It's expected that the feature is turned on only after whole cluster runs with new version, where the feature is already available. Turning on the feature in heterogeneous cluster could lead to non-deterministic behavior in username querying (depending on the fact what server actually serves the request)

**Homogeneous cluster**
It may happen there are still some user entities in the store with entity version older than the current one. This may lead to a situation when the feature is enabled for the realm and new user will be created within the realm differing only by letter case. In this situation `getUserByUsername` should return always the same user and it should be logged a warning. This should be resolved by a mechanism which will allow administrator to see a histogram of versions of entities and to explicitly perform background task migration of objects in the store. Details: https://github.com/keycloak/keycloak-community/blob/main/design/keycloak.x/storage-persistence.md#adding-a-searchable-attribute

Also it's expected that administrator is responsible there are no-collinding usernames upon change of the setting from `case-sensitive` to `case-insensitive` or vice-versa.

**Reply by hmlnarik** [reply]
I agree up to this point:

> [...] resolved by a mechanism which will allow administrator to see a histogram of versions of entities [...]

Entity versions should be transparent to the administrator - the administrator would only be notified that migration is needed and decides when to perform it. Depends on background tasks feature.

**Comment by mariocozta** [comment]
Sorry if this is not the place to ask, but Is there any new version where you can save usernames with upper case characters?

**Reply by vramik** [reply]
Hello @mariocozta, the feature is available since keycloak 20 and only for map store, which is currently experimental feature. It can be turned on by property: https://github.com/keycloak/keycloak/pull/13183/files#diff-727c72cf577645ce6b64fd87bb4d5e900113858b705a9f64875d1a0aa5f54026R43

AFAIK there is no plan to introduce the feature to legacy store.

**Reply by codespearhead** [reply]
This issue seems to have been fixed in the Map Store (an experimental feature) in Keycloak 20 [1][2][3][4][5]. However, since that feature was removed in Keycloak 23 [6][7], the fix should be ported to the JPA store.

There seems to be interest around the port in [8].

***

[1] https://github.com/keycloak/keycloak/discussions/10198
[2] https://github.com/keycloak/keycloak/discussions/10198#discussioncomment-6903879
[3] https://github.com/keycloak/keycloak/issues/10245
[4] https://github.com/keycloak/keycloak/pull/10332
[5] https://github.com/keycloak/keycloak/pull/13183
[6] https://www.keycloak.org/docs/latest/release_notes/index.html#removal-of-the-map-store
[7] https://github.com/keycloak/keycloak/blob/7d8ff710c20ebf2b668b1761dbecbe4bf23f6f80/docs/documentation/upgrading/topics/changes/changes-23_0_0.adoc?plain=1#L183
[8] https://github.com/keycloak/keycloak/issues/32869

**Comment by zelytra** [comment]
Hello ! Is theire now a simple way to retrieve/enable username case sensitive login/retrieve ? I dont see anything about this in the documentation. All issues mention user cache disable, but has i see in this conversation, do you have implement this feature ?

**Reply by vramik** [reply]
Hello @zelytra, 

due to unfortunate decision to discontinue Map store[1], which provided this capability, there is currently no way to store case sensitive usernames and I'm not aware of any plans to introduce it to current store.

[1] https://www.keycloak.org/2023/10/map-store-removal.html

**Reply by zelytra** [reply]
No official solution okay, but is their any bypass to handle username case ? (Rollback keycloak version, custom mapper or input on register ?)

**Reply by vramik** [reply]
The current store automatically lowers the case when the user is being created: https://github.com/keycloak/keycloak/blob/1d8744e6c1dc99e9dce9d460748efb10a317a1f5/model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java#L116

What you can try to do is to implement you own User Storage SPI: https://www.keycloak.org/docs/latest/server_development/index.html#_user-storage-spi but I do not foresee all the implications, e.g. if Keycloak doesn't expect lowered case username at some places

**Reply by zelytra** [reply]
Alright thank you for your answer !

**Reply by AnubisFUp** [reply]
@vramik Hello!
Just when developing the driver, I encountered the same problem. Can you tell me which class or method needs to be implemented in the driver to redefine the login register when it is received by the driver after being sent by the user?

BTW, the user fedaration driver is used for read-only mode with Augmenting.
https://www.keycloak.org/docs/latest/server_development/index.html#augmenting-external-storage
